datafusion/functions/src/math/round.rs

```diff
@@ -128,29 +134,41 @@ pub fn round(args: &[ArrayRef]) -> Result<ArrayRef> {
                     }
                 )) as ArrayRef)
             }
-            ColumnarValue::Array(decimal_places) => Ok(Arc::new(make_function_inputs2!(
-                &args[0],
-                decimal_places,
-                "value",
-                "decimal_places",
-                Float64Array,
-                Int64Array,
-                {
-                    |value: f64, decimal_places: i64| {
-                        (value * 10.0_f64.powi(decimal_places.try_into().unwrap()))
-                            .round()
-                            / 10.0_f64.powi(decimal_places.try_into().unwrap())
+            ColumnarValue::Array(_) => {
+                let ColumnarValue::Array(decimal_places) =
+                    decimal_places.cast_to(&Int32, None).map_err(|e| {
+                        exec_datafusion_err!("Invalid values for decimal places: {e}")
+                    })?
+                else {
+                    panic!("Unexpected result of ColumnarValue::Array.cast")
```

Comment:
I think you can avoid the panic by collaspsing the two `ColumnarValue::Array` 

So instead of

```rust
            ColumnarValue::Array(_) => {
   let ColumnarValue::Array(decimal_places) =
                    decimal_places.cast_to(&Int32, None).map_err(|e| {
                        exec_datafusion_err!("Invalid values for decimal places: {e}")
                    })?
                else {
..
```

```rust
  ColumnarValue::Array(decimal_places) => { 
...
```

Reply:
but I need to squeeze the `cast_to` in there, right? I didn't find a way how to do that more elegantly

it's the same thing like the `let decimal_places: i32 = decimal_places.try_into()` in the scalar case, except that we can't just `try_into` since the type is dynamic in the arrow array

Reply:
and the panic in the else part is there because we "forget" that the result of the `cast_to` is still an array, ie, it doesnt change into a scalar